### PR TITLE
Display more details about the gas fees

### PR DESCRIPTION
**Pull request by mm-zk**
# What :computer: 
* Added more debugging information on fees
* including the very detailed breakdown (how many come from running the code vs sealing the block)

# Why :hand:
* To improve the understanding about our fees, and start the discussions on what can be improved

# Evidence :camera:
image

# Other:
* this PR adds some changes to the local version of the bootloader - which we'll have to persist for how (and re-apply when we update bootloader code). 
* in the near future (when new version of the VM api arrives with custom tracers) , we'll be able to use the tracer to handle some of these informations instead.
* also it seems that in memory node uses a different amount of gas, when we replay the mainnet transactions - with this PR we have tools to debug it better

### CODE REVIEW

File: src/node.rs [hunk 1]
```diff
@@ -781,6 +785,133 @@ impl<S: ForkSource + std::fmt::Debug> InMemoryNode<S> {
...
+            "Gas - Limit: {} | Used: {} | Refunded: {}",
+            to_human_size(total_gas_limit),
+            to_human_size(gas_used),
+            to_human_size(debug.refund_by_operator)
+        );
+
+        if debug.total_gas_limit_from_user != total_gas_limit {
+            println!(
+                "  WARNING: user actually provided more gas {}, but system had a lower max limit.",
+                to_human_size(debug.total_gas_limit_from_user)
```
Comment (MexicanAce) [inline src/node.rs:825]: Since these are warnings, should we post this message in yellow font for better awareness?

File: src/node.rs [hunk 2]
```diff
@@ -781,6 +785,133 @@ impl<S: ForkSource + std::fmt::Debug> InMemoryNode<S> {
...
+            to_human_size(gas_for_validation * 100 / gas_used)
+        );
+        println!(
+            "  {:>15} gas ({:>2}%) for computations (opcodes)",
+            to_human_size(gas_spent_on_compute),
+            to_human_size(gas_spent_on_compute * 100 / gas_used)
+        );
+
+        println!("\n\n === Transaction setup costs ===");
+        let publish_block_l1_bytes = 58823;
```
Comment (MexicanAce) [inline src/node.rs:877]: Where does `58_823` come from?

File: src/utils.rs [hunk 3]
```diff
@@ -123,3 +123,37 @@ pub fn adjust_l1_gas_price_for_tx(
         l1_gas_price.as_u64()
     }
 }
+
+/// Takes long integers and returns them in human friendly format with "_".
+/// For example: 12_334_093
+pub fn to_human_size(input: U256) -> String {
```
Comment (MexicanAce) [inline src/utils.rs:129]: Thank you for this!!!

File: src/node.rs [hunk 4]
```diff
@@ -841,12 +975,19 @@ impl<S: ForkSource + std::fmt::Debug> InMemoryNode<S> {
...
-            tx.gas_limit() - tx_result.gas_refunded,
-            tx_result.gas_refunded
+            "Gas - Limit: {} | Used: {} | Refunded: {}",
+            to_human_size(tx.gas_limit()),
+            to_human_size(tx.gas_limit() - tx_result.gas_refunded),
+            to_human_size(tx_result.gas_refunded.into())
         );
 
+        if self
+            .display_detailed_gas_info(&vm, spent_on_pubdata)
```
Comment (MexicanAce) [inline src/node.rs:985]: As there's a lot of information printed to the console about gas details, do we want to have a flag similar to `show-calls`?

**Comment by idea404** [comment]
**On the screenshot**

It shows how much gas was used in publishing to L1 (a figure of 1.7mil gas), but from the table overview below I can not see how there is a subset of rows whose totals sum to that value which I am finding difficult to understand. 

Then we see another table below showing the trx setup costs with numbers that can not be seen anywhere in the table we just saw above up until we see the total trx setup cost which matches the trx setup row for the table above. I'm left wondering why are the other numbers shown if there is no relationship between them and the transaction expense for this transaction (ie the rows from the table below).

File: src/utils.rs (hunk 3 above)
Comment (mm-zk) [reply src/utils.rs:129]: sure thing.l

**Review by MexicanAce — APPROVED** [review]
LGTM

**Pull request merged**